homeassistant/components/updater.py

```diff
@@ -0,0 +1,143 @@
+    repo_branch = config['updater'].get(CONF_REPO_BRANCH, 'master')
+
+    # create updater service
+    def run_update(service_call):
+        ''' run the updater '''
+        upscript = os.path.join(HA_SOURCE_DIR, 'scripts', 'update')
+        mypid = str(os.getpid())
+        cmd = [upscript, HA_SOURCE_DIR, mypid]
+
+        if log_file is not None:
+            cmd.append(log_file)
+            if pid_file is not None:
+                cmd.append(pid_file)
+
+        subprocess.call(cmd)
+    hass.services.register(DOMAIN, 'update', run_update)
+
+    # find local copy info
+    try:
+        from git import Repo
+    except ImportError:
+        _LOGGER.error('Missing package gitpython')
+        current_sha = None
+    else:
+        repo = Repo(HA_SOURCE_DIR)
+        if repo.bare:
+            current_sha = None
+        else:
+            current_sha = repo.head.commit.hexsha
+
+    # create component entity
+    Updater(hass, current_sha, repo_name, repo_branch, pid_file, log_file)
+
+    return True
+
+
+class Updater(Entity):
+    """ Updater entity class """
+
+    hass = None
+    entity_id = 'updater.updater'
+    name = 'Updater'
+
+    def __init__(self, hass, current_sha, repo_name, branch, pid_file,
+                 log_file):
+        # pylint: disable=too-many-arguments
+        self.hass = hass
+        self.config = {'repo_name': repo_name, 'branch': branch,
+                       'pid_file': pid_file, 'log_file': log_file}
+
+        # register update check event
+        def check_ha_update(event):
+            """ Check GitHub for avaialable update """
+            self.update()
+        hass.track_time_change(check_ha_update, hour=0, minute=0, second=0)
+
+        # initialize data
+        self.versions = {'newest_sha': None, 'newest_msg': None,
+                         'newest_date': None, 'newest_url': None,
+                         'current_sha': current_sha}
+
+        self.update()
+
+    def update(self):
+        ''' Update the state of the entity '''
+        _LOGGER.info('Looking for updates.')
+
+        # pull data from github
+        github_resp = requests.get(
+            GH_API_CALL.format(repo=self.config['repo_name'],
+                               branch=self.config['branch']))
+        github_data = github_resp.json()
+        if github_resp.headers['status'] == '200 OK':
+            self.versions['newest_sha'] = github_data[0]['sha']
+            self.versions['newest_msg'] = github_data[0]['commit']['message']
+            self.versions['newest_date'] = \
+                github_data[0]['commit']['author']['date']
+            self.versions['newest_url'] = github_data[0]['html_url']
+        else:
+            _LOGGER.warning('Error communicating with GitHub')
+
+        # update with HA
+        self.update_ha_state()
+
+    @property
+    def hidden(self):
+        ''' only show the entity when an update is available '''
+        return not self.versions['newest_sha'] and \
```

Comment:
If you fail the setup of the component if `current_sha` cannot be detected, you can simplify this check with `return self.newest_sha != self.current_sha`